### PR TITLE
[tensorparallel] Updated TP layers

### PR DESCRIPTION
**Pull request by kurisusnowdeng**
- Polished tensor parallel implementation
- Achieved the same 1D throughput as the latest Megatron

### CODE REVIEW

**Comment by feifeibear** [comment]
Please add some unitests!

**Comment by kurisusnowdeng** [comment]
> Please add some unitests!

We already have tests for tensor parallel layers

File: colossalai/context/parallel_mode.py [hunk 1]
```diff
@@ -39,6 +39,8 @@ class ParallelMode(Enum):
     PARALLEL_3D_INPUT = '3d_input'
     PARALLEL_3D_WEIGHT = '3d_weight'
     PARALLEL_3D_OUTPUT = '3d_output'
+    PARALLEL_3D_INPUT_X_WEIGHT = "3d_input_x_weight"
+    PARALLEL_3D_OUTPUT_X_WEIGHT = "3d_output_X_weight"
```
Comment (feifeibear) [inline colossalai/context/parallel_mode.py:43]: why 3d_output_**X** uppercase?

File: colossalai/context/process_group_initializer/initializer_3d.py [hunk 2]
```diff
@@ -176,6 +176,112 @@ def init_dist_group(self):
...
+            Tuple (local_rank, group_world_size, process_group, ranks_in_group, mode):
+                3D tensor parallelism's information among input in a tuple.
+        """
+        local_rank = None
+        ranks_in_group = None
+        process_group = None
+        cpu_group = None
+        group_world_size = None
+        mode = ParallelMode.PARALLEL_3D_INPUT_X_WEIGHT
+        env.input_x_weight_group_3d = mode
```
Comment (feifeibear) [inline colossalai/context/process_group_initializer/initializer_3d.py:211]: is env a global var? If so, it should be in uppercase.

File: colossalai/nn/layer/parallel_1d/_operation.py [hunk 3]
```diff
@@ -43,3 +45,52 @@ def backward(ctx, grad_output):
...
+        total_input = total_input.view(total_input.shape[0] * total_input.shape[1], total_input.shape[2])
+
+        if ctx.async_grad_allreduce:
+            # Asynchronous all-reduce
+            handle = dist.all_reduce(grad_input, group=gpc.get_group(ctx.parallel_mode), async_op=True)
+            # Delay the start of weight gradient computation shortly (3us) to have
+            # all-reduce scheduled first and have GPU resources allocated
+            _ = torch.empty(1, device=grad_output.device) + 1
+
+        grad_weight = grad_output.t().matmul(total_input)
```
Comment (feifeibear) [inline colossalai/nn/layer/parallel_1d/_operation.py:86]: Can we use total_input here? Isn't it being allreduced?

File: colossalai/context/process_group_initializer/initializer_3d.py (hunk 2 above)
Comment (kurisusnowdeng) [reply colossalai/context/process_group_initializer/initializer_3d.py:211]: it is a global variable manager

File: colossalai/nn/layer/parallel_1d/_operation.py (hunk 3 above)
Comment (kurisusnowdeng) [reply colossalai/nn/layer/parallel_1d/_operation.py:86]: 1d weights do not need to be allreduced

Comment (feifeibear) [reply colossalai/nn/layer/parallel_1d/_operation.py:86]: I see

**Review by feifeibear — APPROVED** [review]

File: colossalai/context/parallel_mode.py (hunk 1 above)
Comment (kurisusnowdeng) [reply colossalai/context/parallel_mode.py]: fixed

**Pull request merged**